### PR TITLE
Disallow null/empty or duplicate composite sources

### LINKED ISSUE #32414: Composite aggregation is inconsistent
When defining `sources` for `composite` aggregation it expects a list of objects where each object has a single key - the name of the source - and value - the actual aggregation. This is super unexpected pattern as in other places where the DSL has similar requirements it just accepts an object with multiple keys - names of the subaggregations (like `aggs` itself for example or a list of aggregations directly without a name if order is important (like filters agg with anonymous filters).

Additionally there is no protection against duplicate names of sources so a following example just uses the second aggregation, silently ignoring the first definition:

```
curl localhost:9200/_search -d '{
  "aggs": {
    "my": {
      "composite": {
        "size": 1,
        "sources": [
          {"test": {"terms": {"field": "files"}}},
          {"test": {"terms": {"field": "stats.lines"}}}
        ]
      }
    }
  }
}'
```

Could we perhaps do the same as with `filters` for consistency? Use keyed request if order does not matter (order can be either implicit by key or explicit by specifying `order` key at the top level) or a list of anonymous objects? If not, it would still be great to have protection against duplicate keys - the above example should raise an exception to alert the user they are doing something unsupported.

Thank you!

### PR DESCRIPTION
**Pull request by polyfractal**
Adds some validation to prevent duplicate source names from being used in the composite agg, closes https://github.com/elastic/elasticsearch/issues/32414.

While adding that I ran into an NPE due to how the parser uses a private ctor that passes in a null source (and then sets it during parsing).  I refactored to use a `ConstructingObjectParser` and removed the private ctor and setter for sources, making it mandatory.  

`ConstructingObjectParsers` are a bit clunky for named objects like aggs, but it seemed better than putting null checks on all the places that could set the sources.  I can do that if preferred though

Also updated the docs to state that source order is important.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilder.java [hunk 1]
```diff
@@ -45,35 +48,43 @@
...
     public CompositeAggregationBuilder(String name, List<CompositeValuesSourceBuilder<?>> sources) {
         super(name);
+        validateSources(sources);
         this.sources = sources;
     }
 
     protected CompositeAggregationBuilder(CompositeAggregationBuilder clone,
                                           AggregatorFactories.Builder factoriesBuilder, Map<String, Object> metaData) {
         super(clone, factoriesBuilder, metaData);
+        validateSources(clone.sources);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilder.java]: nit: the original source are already validated so maybe no need to do it twice

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilder.java [hunk 2]
```diff
@@ -167,6 +173,25 @@ public int size() {
...
     }
 
+    private static void validateSources(List<CompositeValuesSourceBuilder<?>> sources) {
+        if (sources == null || sources.isEmpty()) {
+            throw new IllegalArgumentException("Composite [" + SOURCES_FIELD_NAME.getPreferredName() + "] cannot be null or empty");
+        }
+
+        Set<String> names = new HashSet<>();
+        Set<String> duplicates = new HashSet<>();
+        sources.forEach(source -> {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilder.java]: should it also check that the inner sources are not null ?

**Review by jimczi — APPROVED** [review]
Thanks @polyfractal , it looks good to me. I left some minor comments but no need for another review.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilder.java (hunk 1 above)
Comment (polyfractal) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilder.java]: whoops good catch

**Pull request merged**